### PR TITLE
Add Content-Disposition http header

### PR DESCRIPTION
**Pull request by nizsheanez**
For https://github.com/ethersphere/go-ethereum/issues/527

### CODE REVIEW

File: swarm/api/filesystem.go
Comment (acud) [inline swarm/api/filesystem.go:107]: you have removed important I/O throttling code here. either convert to a worker pattern or preserve as is

File: cmd/swarm/upload.go
Comment (acud) [inline cmd/swarm/upload.go]: i wonder if we really need to open a new file handle inside of `api.DetectContentType`. at this point we already have `f` as an open file handle. maybe we could use that instead. WDYT?

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: comment should follow godoc style:
```
// DetectContentType by file content with fallback
// to file extension and "application/octet-stream"
```

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: extract to `const MIME_OCTET_STREAM = "application/octet-stream"`, then reuse across the code (DRY)

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: `log.Error` - err != nil
also - you're not returning on error (the rest of the code below will continue executing - resulting in an error). if you choose to have another handle opened here - please return an empty string in this code block - the function has to return here.
as mentioned above - i would remove this part and already use the open handle that is created in the enclosing call

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: comment not necessary - code is self explanatory :)

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: use the `const` from above

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: use `const`

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: this entire function can sit inside `server.go` as it has no `api` context and is used exclusively within the `http` package

File: swarm/api/content_type.go
Comment (acud) [inline swarm/api/content_type.go]: i don't think that there's justification for a new file here. split the relevant parts into `api.go` and `server.go` (see below)

File: swarm/api/content_type_test.go
Comment (acud) [inline swarm/api/content_type_test.go]: same with this file - please remove and split into `server_test.go` and `api_test.go`

File: swarm/api/content_type_test.go
Comment (acud) [inline swarm/api/content_type_test.go]: use `ioutil.TempFile` instead

File: swarm/api/content_type_test.go
Comment (acud) [inline swarm/api/content_type_test.go]: `content` param not used - can be removed

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go]: two issues with this:
1. do not use r.URL as it will contain the entire requested URI e.g. `/bzz:/somename.eth/something` etc. please use the `uri` variable above which comes from the middleware that parses the bzz uri. it also has a `Path` member which you can use to get to the path inside the bzz uri.
2. there could be a situation where you would be downloading a tarball but the filename might be empty. in that case, you should fallback to the `Addr` from `uri` (or `untitled`)

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go]: use `const`

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go]: same as above

**Review by acud — CHANGES_REQUESTED** [review]
Thanks for the contribution 👍 🍺 

Please see my comments in the code.

File: cmd/swarm/upload.go
Comment (nizsheanez) [reply cmd/swarm/upload.go]: I found problem: 
- If DetectContentType func will accept *File, then it must current preserve position in file
- Here we working with `*swarm.File` not `*os.File` which has no Seek() method to save and restore position

It doesn't look possible to get content probe from *swarm.File 
Maybe you have an idea?

File: swarm/api/filesystem.go
Comment (nizsheanez) [reply swarm/api/filesystem.go:107]: Yay! I will revert it.

File: cmd/swarm/upload.go
Comment (nizsheanez) [reply cmd/swarm/upload.go]: Ah, *swarm.File it's local file, then i can cast it to Seeker

Comment (acud) [reply cmd/swarm/upload.go]: you can also try to change the `swarm.File` spec to be
```
Reader,
Closer,
Seeker
```
instead of just `ReadCloser`. that might make your life easier. but you also might need to move the type around a bit because this will probably generate a circular dependency when you'll try to reference `client` from `api` :)

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go]: even if its one file - if there's an enclosing manifest, i think there should be a filename there too (need to double check. in any case - there should be - let's assume that)

**Comment by nizsheanez** [comment]
Fixed all comments, but I need carefully revisit code about " I/O throttling" tomorrow. 

Found edge case: 
if .css file has any content - `http.DetectContentType` return text/plain 
looks like in case of text/plain also need to fallback to extension-based check
Do you feel ok with it? Not sure if other edge cases are around...

**Comment by nizsheanez** [comment]
I will look around to existing webservers tonight. For example Nginx has mime configuration file, default is: https://github.com/nginx/nginx/blob/master/conf/mime.types

For deeper content check https://github.com/h2non/filetype may help, but it looks overkill for me.

**Comment by nizsheanez** [comment]
Here are my thoughts.

About content sniff: 
- (thank https://github.com/ipfs/go-ipfs/issues/2164 ) - Found that `http.DetectContentType` is already used when `http.ServeContent` is invoked. It's mean - need to remove Content-Type detection when we serving content, just double-check that we using `http.ServeContent`. And I will copy-paste code from there and check if it will work well for Uploading (there are extension check firs... i still feel that we can sniff first and fallback when sniff return "text/plain; charset=utf-8" or "application/octet-stream", but too much open questions - `why don't fallback on "text/plain; charset=utf-16be"?` or `is there any other edge cases?` - let's go with ext check first, and maybe open another issue for possible attacks research - because now attacker don't need to put crazy content to .css file, he just can provide any valid Content-Type header and we will trust it). 
- Fun question: html - What if client sent us header "X-Content-Type-Options=nosniff"?
- I saw that swarm/api/testdata has Website serving test, I think need to add non-utf-8 Website to test (but there is a mess :-) .css spec forcing utf-8, while .html and .js allow other encodings :-) ) 

About extension-based checks:
- `mime.TypeByExtension` using not only `builtinTypesLower` but also OS preconfigured list (see `mime.osInitMime`)
- Swarm expecting be run in diverse OS/Platforms/Docker - I advise to hardcode own list of extension-based ContentTypes ( at list copy from https://github.com/nginx/nginx/blob/master/conf/mime.types or https://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types ), just to make Swarm behave similar in different environments. Any cons?

**Comment by acud** [comment]
> need to remove Content-Type detection when we serving content, just double-check that we using http.ServeContent. And I will copy-paste code from there and check if it will work well for Uploading.

We don't use it. When fetching content from Swarm - we usually use `io.Copy` to write into the response, not `http.ServeContent`. See `server.go#HandleBzzGet`

> Fun question: html - What if client sent us header "X-Content-Type-Options=nosniff"?

Ignore? this is a [response header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options).
We should definitely add this header to our responses. But let's first make sure that this works otherwise we could be breaking stuff (the header should be set only if we have a content type set - to fallback when a scenario of no content-type in manifest occurs)

> I saw that swarm/api/testdata has Website serving test, I think need to add non-utf-8 Website to test (but there is a mess :-) .css spec forcing utf-u8, while .html and .js allow other encodings :-) )

Definitely, I also think that adding tests for utf8 content (filenames included) is a must. I'll open an issue for that :)

>  advise to hardcode own list of extension-based ContentTypes

I completely agree. The list from nginx looks pretty exhaustive for this point in time.

**Comment by nizsheanez** [comment]
Ok. Added. Not solved parts yet:
- at `swarm up README.md` client doesn't send file name (even if send, looks like no place to store it... ) - I will investigate tomorrow

**Comment by acud** [comment]
This is due to another problem and is outside the scope of this issue. I'm already have this issue at mind and it will be taken care of as part of ongoing efforts to refactor and cleanup swarm manifests.

File: swarm/api/api.go
Comment (acud) [inline swarm/api/api.go]: not needed - use `init()` instead

File: swarm/api/api.go
Comment (acud) [inline swarm/api/api.go]: rename to `init()`

File: swarm/api/api.go
Comment (acud) [inline swarm/api/api.go]: [not necessary ](https://golang.org/doc/effective_go.html?#init)

File: swarm/api/api_test.go
Comment (acud) [inline swarm/api/api_test.go]: remove comment

File: swarm/api/filesystem.go
Comment (acud) [inline swarm/api/filesystem.go]: I'd rather have this `wait()` before the mime detection happens, as in the original code

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go]: please amend this - use `uri` from context instead of request URI. also - provide the filename fallback

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go]: @lmars what is your input on this?

File: swarm/api/server.go
Comment (acud) [inline swarm/api/server.go]: remove file

**Review by acud — CHANGES_REQUESTED** [review]

**Comment by nizsheanez** [comment]
Fixed all comments. 

Question: do i need to delete force of "application/octet-stream" in `bzz-resource`: https://github.com/ethersphere/go-ethereum/blob/master/swarm/api/http/server.go#L659  ?
Next line `http.ServeContent` must sniff content well.

**Comment by acud** [comment]
Great work on this! :)
the code is fine as-is - `http.ServeHttp` tries to sniff the content just in the case that `Content-Type` [isn't set](https://golang.org/src/net/http/fs.go) (L191-L193)

**Comment by nizsheanez** [comment]
Thank you for on-boarding.

File: swarm/api/api.go
Comment (lmars) [inline swarm/api/api.go]: I personally feel this `mime.types` file should be vendored (or just copied into a Go file) to make it easy to keep in-sync with the upstream file.

I also think it should be clear why this is here (if I saw this I'd probably just think it was unnecessary because the Go mime package does the right thing in pretty much all cases).

File: swarm/api/api.go
Comment (lmars) [inline swarm/api/api.go]: This should panic if there is an error (in case some one adds an extension or mime type that doesn't parse)

File: swarm/api/api.go
Comment (lmars) [inline swarm/api/api.go]: This comment is not consistent with what the method actually does.

File: swarm/api/api_test.go
Comment (lmars) [inline swarm/api/api_test.go]: You could avoid the need for a temp file by modifying `DetectContentType` to take `(filename string, data io.ReadSeeker)` and just use a `bytes.Reader` in the tests.

File: swarm/api/api_test.go
Comment (lmars) [inline swarm/api/api_test.go]: This should use [`t.Run`](https://golang.org/pkg/testing/#T.Run) to isolate each test.

File: swarm/api/http/server.go
Comment (lmars) [inline swarm/api/http/server.go]: Are you sure we want to make all files attachments? If I go to `bzz://mysite.eth/index.html` in my browser, I expect to see the HTML in the browser, not download it to a local file.

File: swarm/api/http/server.go
Comment (lmars) [reply swarm/api/http/server.go]: I'm not really sure what the purpose of this code is?

**Comment by nizsheanez** [comment]
2 more changes: 
- Can't use `ioutil.TempFile` - it returns predictable extension only since go1.11 (see https://github.com/golang/go/issues/4896), solution - a bit of copy-paste (#527, #944)
- Go returns different ContentType for .md files on Debian and OSx (in image golang:1.11 and golang:1.11-alpine it returns "text/markdown", but on OSx - "text/plain"). As i understand - spec about "text/markdown" exists by not officially accepted yet (https://tools.ietf.org/html/rfc7763). I added .md to our list anyway (feel free to deny if you don't like)

File: swarm/api/api_test.go
Comment (nizsheanez) [reply swarm/api/api_test.go]: indeed, it will simplify everything

File: swarm/api/http/server.go
Comment (nizsheanez) [reply swarm/api/http/server.go]: oh, how much you right. Right now i will change it to:
`Content-Disposition: inline; filename="example.html"`
But i will rise question in Gitter - maybe exist some cases when `attachment` make sense, for example in S3 owner can add any headers to it's file/folder by UI, and `Content-Disposition: attachment` too.

**Comment by nizsheanez** [comment]
@lmars about `I'm not really sure what the purpose of this code is?` - I just moved Content-Type header check from 3 places to that function (ValidateContentTypeHeader). Do you feel it unnecessary or need rename to `RequireContentTypeHeader`?

File: swarm/api/api.go
Comment (nizsheanez) [reply swarm/api/api.go]: By vendoring of mime.types file - you mean write some `go generate` command to parse it and gen code? Same approach as here https://github.com/bradrydzewski/go-mimetype (it based on mailcap instead of nginx, but there is same mime file) ?

**Comment by acud** [comment]
@zelig let's first resolve all issues here then squash+reopen on upstream

**Comment by nizsheanez** [comment]
Status: fixed all CR comments except one:
- From Imars: "I personally feel this mime.types file should be vendored (or just copied into a Go file) to make it easy to keep in-sync with the upstream file." - Waiting for hist reply now. 
I have airplane tomorrow, will push something about it - after tomorrow.

File: swarm/api/api.go
Comment (lmars) [reply swarm/api/api.go]: Something like that, yes, which would make it much easier to keep in sync, and more obvious where the types come from.

Comment (acud) [reply swarm/api/api.go]: @lmars do you think the overhead is justified in this case? I mean, the mime types base list doesn't fluctuate [that much](https://github.com/nginx/nginx/commits/master/conf/mime.types).

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go]: validating incoming mime type on upload from header, to make sure that the input we're getting is correct before storing it in a manifest - just to make sure we are on par with IETF RFCs

File: swarm/api/http/server_test.go
Comment (acud) [inline swarm/api/http/server_test.go]: remove comment

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go:209]: i am wondering - maybe in this case we should actually send an `attachment` header. @lmars?

File: swarm/api/http/server.go
Comment (acud) [inline swarm/api/http/server.go:906]: @lmars I think this should be:
if contentType == 'application/*' - return attachment
else return inline.
WDYT?

Comment (nizsheanez) [reply swarm/api/http/server.go:906]: Worry about "application/pdf" and "application/rss+xml", I'm personally fine if .pdf opening in browser (re-checked, browsers do download .pdf file with `Content-Disposition: attachment; file="name.pdf"` header, and only Safari on my Mac auto-open it in Preview).

**Comment by nizsheanez** [comment]
Added generator based on mime.types file as lmars recommended.

File: cmd/swarm/mimegen/generator.go
Comment (lmars) [inline cmd/swarm/mimegen/generator.go:25]: This now seems to do exactly what the package you linked to does (i.e. https://github.com/bradrydzewski/go-mimetype), should we not just use that package?

Comment (nizsheanez) [reply cmd/swarm/mimegen/generator.go:25]: That package looks unmaintained, but let me try to rise PR there, here is list of differences: 
- `go-mimetype` doesn't fail if `mime.AddExtensionType` returns error
- outdated mime.types file

Comment (lmars) [reply cmd/swarm/mimegen/generator.go:25]: Ok, I don't mean for you to go back and forth on this. I assumed you didn't want to use that package because you wanted to use nginx mime types, but if we want to use the same method as that package then we should try and PR upstream (the maintainer looks active on other projects)

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go:906]: i've seen the rss issue too but i think its a non-issue as its a machine readable format and is not supposed to be read by the user as-is (AFAIK)

File: cmd/swarm/mimegen/generator.go
Comment (nizsheanez) [reply cmd/swarm/mimegen/generator.go:25]: Ah, just that mime.types file easier to parse. File from nginx has nginx.config format you can see there are some bracers: https://github.com/nginx/nginx/blob/master/conf/mime.types 

No problem, already created PR: https://github.com/bradrydzewski/go-mimetype/pull/1

Anyway we still discussing about `Content-Disposition`

**Comment by lmars** [comment]
@nizsheanez @justelad it seems this PR is covering two things:

- Improved detection and validation of `Content-Type`
- Add `Content-Disposition`

It seems we've got to a good solution for improving `Content-Type`, but I don't think it's clear what problem we are trying to solve with the `Content-Disposition` header.

What is driving the decision to add this `Content-Disposition` header? Is it a user-requested feature? If it has been requested, I personally think we should expose some form of option to the user to allow them to set `Content-Disposition` if they want, but I don't think Swarm should be trying to preempt how the user wants their content to be served (either inline or as an attachment).

**Comment by acud** [comment]
@lmars this solves the problem that the browser does not always know under which filename to save swarm content and this is rather annoying. 
`Content-Disposition` solves this.
No user requested this. I noticed this problem and opened the ticket on my own accord.

**Comment by nizsheanez** [comment]
I agree that `Content-Disposition: attachment` must be configurable by user - Because we discovered too much edge cases already (maybe even on MRU level instead of file, but i'm not sure). 

Adding `Content-Disposition: inline` - solved `browser does not always know under which filename to save swarm content` problem (with one exception: when client uploading single file - it doesn't send to server fileName. Justelad, says that it's out of scope and he will fix it later: https://github.com/ethersphere/go-ethereum/pull/945#issuecomment-423797672 )

@justelad , now I need to do squash and open new PR with good description?

File: swarm/api/http/server.go
Comment (acud) [reply swarm/api/http/server.go:906]: ok - keep it as it is for now (`inline` it)

**Comment by acud** [comment]
@nizsheanez many thanks for this comprehensive PR!

Please squash all commits and reopen this PR against `ethereum/master` with a description.
:shipit:

**Review by acud — APPROVED** [review]

**Comment by nizsheanez** [comment]
Done https://github.com/ethereum/go-ethereum/pull/17782

**Pull request closed without merging**
